In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
# 데이터 불러오기
path='../data/Paris/listings.csv'
pd.read_csv(path)

id                                       listing_url  \
0                     3109                 https://www.airbnb.com/rooms/3109   
1                   165409               https://www.airbnb.com/rooms/165409   
2                     5396                 https://www.airbnb.com/rooms/5396   
3                   166370               https://www.airbnb.com/rooms/166370   
4                    39948                https://www.airbnb.com/rooms/39948   
...                    ...                                               ...   
84392  1112562855977040914  https://www.airbnb.com/rooms/1112562855977040914   
84393  1112591023192990750  https://www.airbnb.com/rooms/1112591023192990750   
84394  1112842648117310093  https://www.airbnb.com/rooms/1112842648117310093   
84395  1112843324939236177  https://www.airbnb.com/rooms/1112843324939236177   
84396  1112849968490989659  https://www.airbnb.com/rooms/1112849968490989659   

            scrape_id last_scraped           source  \
0      20240316023852   2024-03-19      city scrape   
1      20240316023852   2024-03-19      city scrape   
2      20240316023852   2024-03-18      city scrape   
3      20240316023852   2024-03-16      city scrape   
4      20240316023852   2024-03-19  previous scrape   
...               ...          ...              ...   
84392  20240316023852   2024-03-19      city scrape   
84393  20240316023852   2024-03-18      city scrape   
84394  20240316023852   2024-03-19      city scrape   
84395  20240316023852   2024-03-18      city scrape   
84396  20240316023852   2024-03-18      city scrape   

                                                    name  \
0                                           zen and calm   
1                        studio charme Haut Marais PARIS   
2           Your perfect Paris studio on Île Saint-Louis   
3                      Big appartment with view 4 people   
4                 COSY INDEPENDENT ROOM IN CENTRAL PARIS   
...                                                  ...   
84392  Artists neighbour apt /private yard 10mn Print...   
84393  Appartement Tour Eiffel - Porte Dauphine - 2 Pers   
84394       1621: Prime spot in Paris Olympic Games 2024   
84395              Stylish flat in the heart of the city   
84396   1637: Prime location in Paris Olympic Games 2024   

                                             description  \
0      Lovely Appartment with one bedroom with a Quee...   
1                You'll love this typical parisian flat.   
2      NEW SOFA-BED SINCE JUNE 2023, Please disregard...   
3      2 Bedrooms apartment for up to 4 people, with ...   
4      Large newly-renovated guest room in central Pa...   
...                                                  ...   
84392  Simplify your life at this peaceful place<br /...   
84393  Nestled in the heart of the prestigious 16th a...   
84394  Discover the allure of our generously proporti...   
84395  Welcome to Paris! We are delighted to welcome ...   
84396  Are you yearning for a romantic or adventurous...   

                                   neighborhood_overview  \
0      Good restaurants<br />very close the Montparna...   
1                                                    NaN   
2      You are within walking distance to the Louvre,...   
3      The avenue is located nearby the animated dist...   
4      Located in the heart of Paris, The Montorgueil...   
...                                                  ...   
84392                                                NaN   
84393  A l’orée du bois, le jardin du Ranelagh est un...   
84394  Embark on an enchanting voyage of discovery th...   
84395  The 6th arrondissement is home to Saint-Germai...   
84396  Nestled within the vibrant heart of the 14th A...   

                                             picture_url    host_id  ...  \
0      https://a0.muscache.com/pictures/miso/Hosting-...       3631  ...   
1      https://a0.muscache.com/pictures/miso/Hosting-...     788732  ...   
2      http

In [3]:
raw = pd.read_csv(path)
df = raw.copy()

회의 결과대로 컬럼 나누기
 - 숙소 테이블
 - 리뷰 테이블
 - 선별할 컬럼 테이블

숙소 테이블 만들기

In [4]:
df_lodging = df[['id', 'host_id','host_response_time','host_response_rate','host_acceptance_rate','host_is_superhost',
                 'host_total_listings_count','neighbourhood_cleansed','room_type','accommodates','bathrooms','bedrooms',
                 'beds','amenities','price','minimum_nights','maximum_nights']]
df_lodging.head(3)

id  host_id  host_response_time host_response_rate  \
0    3109     3631        within a day               100%   
1  165409   788732  within a few hours               100%   
2    5396     7903      within an hour               100%   

  host_acceptance_rate host_is_superhost  host_total_listings_count  \
0                  80%                 f                        2.0   
1                  96%                 t                        1.0   
2                  97%                 f                        3.0   

  neighbourhood_cleansed        room_type  accommodates  bathrooms  bedrooms  \
0           Observatoire  Entire home/apt             2        1.0       1.0   
1                 Temple  Entire home/apt             2        1.0       0.0   
2         Hôtel-de-Ville  Entire home/apt             2        1.0       0.0   

   beds                                          amenities    price  \
0   1.0  ["Cooking basics", "Luggage dropoff allowed", ...  $250.00   
1   1.0  ["Lockbox", "Wifi", "Kitchen", "Refrigerator",...  $140.00   
2   1.0  ["Cooking basics", "Long term stays allowed", ...  $107.00   

   minimum_nights  maximum_nights  
0               2              30  
1               3             365  
2               1            1125

In [5]:
# 컬럼명 변경

df_lodging = df_lodging.rename(columns={
    'id': '숙소_id',
    'host_id': '호스트_id',
    'host_response_time': '답변_평균시간',
    'host_response_rate': '문의_응답률',
    'host_acceptance_rate': '예약_수락률',
    'host_is_superhost': '슈퍼호스트',
    'host_total_listings_count': '숙소_수',
    'neighbourhood_cleansed': '숙소_지역',
    'room_type': '숙소_유형',
    'accommodates': '수용인원수',
    'bathrooms': '욕실수',
    'bedrooms': '침실수',
    'beds': '침대수',
    'amenities': '편의시설',
    'price': '숙소_가격',
    'minimum_nights': '최소_숙박일',
    'maximum_nights': '최대_숙박일'
})

df_lodging.head(3)


숙소_id  호스트_id             답변_평균시간 문의_응답률 예약_수락률 슈퍼호스트  숙소_수  \
0    3109    3631        within a day   100%    80%     f   2.0   
1  165409  788732  within a few hours   100%    96%     t   1.0   
2    5396    7903      within an hour   100%    97%     f   3.0   

            숙소_지역            숙소_유형  수용인원수  욕실수  침실수  침대수  \
0    Observatoire  Entire home/apt      2  1.0  1.0  1.0   
1          Temple  Entire home/apt      2  1.0  0.0  1.0   
2  Hôtel-de-Ville  Entire home/apt      2  1.0  0.0  1.0   

                                                편의시설    숙소_가격  최소_숙박일  최대_숙박일  
0  ["Cooking basics", "Luggage dropoff allowed", ...  $250.00       2      30  
1  ["Lockbox", "Wifi", "Kitchen", "Refrigerator",...  $140.00       3     365  
2  ["Cooking basics", "Long term stays allowed", ...  $107.00       1    1125

In [6]:
df_lodging.shape

(84397, 17)

리뷰 테이블 만들기

In [7]:
df_review = df[['number_of_reviews', 'number_of_reviews_l30d', 'review_scores_rating', 'review_scores_accuracy', 
                'review_scores_cleanliness', 'review_scores_checkin', 'review_scores_communication', 
                'review_scores_location', 'review_scores_value', 'reviews_per_month']]
df_review.head(3)


number_of_reviews  number_of_reviews_l30d  review_scores_rating  \
0                  4                       0                  5.00   
1                106                       0                  4.68   
2                393                       5                  4.60   

   review_scores_accuracy  review_scores_cleanliness  review_scores_checkin  \
0                    5.00                       5.00                   5.00   
1                    4.72                       4.59                   4.87   
2                    4.63                       4.57                   4.81   

   review_scores_communication  review_scores_location  review_scores_value  \
0                         5.00                    5.00                 5.00   
1                         4.89                    4.95                 4.74   
2                         4.84                    4.96                 4.58   

   reviews_per_month  
0               0.05  
1               0.69  
2               2.19

In [8]:
# 컬럼명 변경
df_review = df_review.rename(columns={
    'number_of_reviews': '리뷰수',
    'number_of_reviews_l30d': '30일_리뷰수',
    'review_scores_rating': '리뷰점수',
    'review_scores_accuracy': '숙소_정확성_리뷰점수',
    'review_scores_cleanliness': '숙소_청결도_리뷰점수',
    'review_scores_checkin': '숙소_체크인_리뷰점수',
    'review_scores_communication': '숙소_소통_리뷰점수',
    'review_scores_location': '숙소_위치_리뷰점수',
    'review_scores_value': '숙소_가격_리뷰점수',
    'reviews_per_month': '평균_리뷰수'
})
df_review.head()


리뷰수  30일_리뷰수  리뷰점수  숙소_정확성_리뷰점수  숙소_청결도_리뷰점수  숙소_체크인_리뷰점수  숙소_소통_리뷰점수  \
0    4        0  5.00         5.00         5.00         5.00        5.00   
1  106        0  4.68         4.72         4.59         4.87        4.89   
2  393        5  4.60         4.63         4.57         4.81        4.84   
3   78        0  4.56         4.74         4.37         4.82        4.79   
4  126        3  4.86         4.93         4.82         4.96        4.97   

   숙소_위치_리뷰점수  숙소_가격_리뷰점수  평균_리뷰수  
0        5.00        5.00    0.05  
1        4.95        4.74    0.69  
2        4.96        4.58    2.19  
3        4.65        4.48    0.51  
4        4.94        4.82    0.99

In [9]:
df_review.shape

(84397, 10)

보류 검증용 컬럼 테이블

In [10]:
df_check = df[['listing_url', 'description', 'host_has_profile_pic', 'host_identity_verified', 'latitude', 
              'longitude', 'has_availability']]
df_check.head()


listing_url  \
0    https://www.airbnb.com/rooms/3109   
1  https://www.airbnb.com/rooms/165409   
2    https://www.airbnb.com/rooms/5396   
3  https://www.airbnb.com/rooms/166370   
4   https://www.airbnb.com/rooms/39948   

                                         description host_has_profile_pic  \
0  Lovely Appartment with one bedroom with a Quee...                    t   
1            You'll love this typical parisian flat.                    t   
2  NEW SOFA-BED SINCE JUNE 2023, Please disregard...                    t   
3  2 Bedrooms apartment for up to 4 people, with ...                    t   
4  Large newly-renovated guest room in central Pa...                    t   

  host_identity_verified  latitude  longitude has_availability  
0                      f  48.83191    2.31870                t  
1                      t  48.86652    2.36261                t  
2                      t  48.85247    2.35835                t  
3                      t  48.87623    2.36869                t  
4                      t  48.87068    2.34460                t

In [11]:
# 컬럼명 변경 
df_check = df_check.rename(columns={
    'listing_url': '숙소_url',
    'description': '숙소_상세설명',
    'host_has_profile_pic': '프로필사진_소지_여부',
    'host_identity_verified': '신원_검증_여부',
    'latitude': '위도',
    'longitude': '경도',
    'has_availability': '숙소_예약가능_여부'
})
df_check.head()

숙소_url  \
0    https://www.airbnb.com/rooms/3109   
1  https://www.airbnb.com/rooms/165409   
2    https://www.airbnb.com/rooms/5396   
3  https://www.airbnb.com/rooms/166370   
4   https://www.airbnb.com/rooms/39948   

                                             숙소_상세설명 프로필사진_소지_여부 신원_검증_여부  \
0  Lovely Appartment with one bedroom with a Quee...           t        f   
1            You'll love this typical parisian flat.           t        t   
2  NEW SOFA-BED SINCE JUNE 2023, Please disregard...           t        t   
3  2 Bedrooms apartment for up to 4 people, with ...           t        t   
4  Large newly-renovated guest room in central Pa...           t        t   

         위도       경도 숙소_예약가능_여부  
0  48.83191  2.31870          t  
1  48.86652  2.36261          t  
2  48.85247  2.35835          t  
3  48.87623  2.36869          t  
4  48.87068  2.34460          t

각 테이블의 정보 확인

In [12]:
# 숙소 테이블 정보 확인
df_lodging.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84397 entries, 0 to 84396
Data columns (total 17 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   숙소_id    84397 non-null  int64  
 1   호스트_id   84397 non-null  int64  
 2   답변_평균시간  48310 non-null  object 
 3   문의_응답률   48310 non-null  object 
 4   예약_수락률   56521 non-null  object 
 5   슈퍼호스트    84306 non-null  object 
 6   숙소_수     84390 non-null  float64
 7   숙소_지역    84397 non-null  object 
 8   숙소_유형    84397 non-null  object 
 9   수용인원수    84397 non-null  int64  
 10  욕실수      60959 non-null  float64
 11  침실수      77480 non-null  float64
 12  침대수      60828 non-null  float64
 13  편의시설     84397 non-null  object 
 14  숙소_가격    60917 non-null  object 
 15  최소_숙박일   84397 non-null  int64  
 16  최대_숙박일   84397 non-null  int64  
dtypes: float64(4), int64(5), object(8)
memory usage: 10.9+ MB


In [13]:
df_lodging['숙소_유형'].unique()

array(['Entire home/apt', 'Private room', 'Shared room', 'Hotel room'],
      dtype=object)

In [14]:
df_lodging['숙소_지역'].unique()

array(['Observatoire', 'Temple', 'Hôtel-de-Ville', 'Entrepôt', 'Bourse',
       'Popincourt', 'Louvre', 'Buttes-Montmartre', 'Buttes-Chaumont',
       'Opéra', 'Batignolles-Monceau', 'Passy', 'Élysée', 'Vaugirard',
       'Ménilmontant', 'Panthéon', 'Gobelins', 'Reuilly', 'Luxembourg',
       'Palais-Bourbon'], dtype=object)

In [15]:
df_lodging['숙소_지역'].nunique()

20

In [16]:
# 결측치 개수 확인
df_lodging.isnull().sum()

숙소_id          0
호스트_id         0
답변_평균시간    36087
문의_응답률     36087
예약_수락률     27876
슈퍼호스트         91
숙소_수           7
숙소_지역          0
숙소_유형          0
수용인원수          0
욕실수        23438
침실수         6917
침대수        23569
편의시설           0
숙소_가격      23480
최소_숙박일         0
최대_숙박일         0
dtype: int64

In [17]:
# 리뷰 태이블 확인
df_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84397 entries, 0 to 84396
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   리뷰수          84397 non-null  int64  
 1   30일_리뷰수      84397 non-null  int64  
 2   리뷰점수         58827 non-null  float64
 3   숙소_정확성_리뷰점수  58793 non-null  float64
 4   숙소_청결도_리뷰점수  58797 non-null  float64
 5   숙소_체크인_리뷰점수  58784 non-null  float64
 6   숙소_소통_리뷰점수   58795 non-null  float64
 7   숙소_위치_리뷰점수   58787 non-null  float64
 8   숙소_가격_리뷰점수   58782 non-null  float64
 9   평균_리뷰수       58822 non-null  float64
dtypes: float64(8), int64(2)
memory usage: 6.4 MB


In [18]:
# 결측치 개수 확인
df_review.isnull().sum()

리뷰수                0
30일_리뷰수            0
리뷰점수           25570
숙소_정확성_리뷰점수    25604
숙소_청결도_리뷰점수    25600
숙소_체크인_리뷰점수    25613
숙소_소통_리뷰점수     25602
숙소_위치_리뷰점수     25610
숙소_가격_리뷰점수     25615
평균_리뷰수         25575
dtype: int64

In [19]:
# 숙소 테이블 결측치 처리를 위한 각 속성 확인
df_lodging.isnull().sum()

숙소_id          0
호스트_id         0
답변_평균시간    36087
문의_응답률     36087
예약_수락률     27876
슈퍼호스트         91
숙소_수           7
숙소_지역          0
숙소_유형          0
수용인원수          0
욕실수        23438
침실수         6917
침대수        23569
편의시설           0
숙소_가격      23480
최소_숙박일         0
최대_숙박일         0
dtype: int64

In [20]:
# 답변_평균시간부터 처리
df_lodging['답변_평균시간'].unique()

array(['within a day', 'within a few hours', 'within an hour', nan,
       'a few days or more'], dtype=object)

In [21]:
# 범주형으로 변환
df_lodging['답변_평균시간'] = df_lodging['답변_평균시간'].astype('category')

In [22]:
df_lodging['답변_평균시간'].value_counts()

답변_평균시간
within an hour        29746
within a few hours     9554
within a day           7087
a few days or more     1923
Name: count, dtype: int64

In [23]:
# 문의_응답률 처리
df_lodging['문의_응답률'].unique()

array(['100%', nan, '50%', '93%', '75%', '70%', '86%', '90%', '80%', '0%',
       '39%', '91%', '33%', '98%', '71%', '67%', '25%', '99%', '72%',
       '40%', '83%', '92%', '85%', '57%', '95%', '77%', '62%', '89%',
       '63%', '84%', '13%', '78%', '43%', '14%', '82%', '88%', '96%',
       '60%', '87%', '11%', '20%', '94%', '44%', '97%', '69%', '30%',
       '29%', '81%', '55%', '64%', '79%', '56%', '17%', '18%', '10%',
       '22%', '38%', '8%', '73%', '3%', '31%', '23%', '6%', '52%', '19%',
       '36%', '58%', '54%', '47%', '4%', '68%', '61%', '59%', '65%',
       '21%', '27%', '76%', '7%', '35%', '53%', '51%', '9%', '15%', '34%',
       '46%', '37%', '16%', '26%', '42%', '12%', '5%', '28%', '45%',
       '49%', '48%', '74%', '2%'], dtype=object)

In [24]:
df_lodging['문의_응답률'].value_counts()

문의_응답률
100%    37075
90%      1197
0%       1176
98%      1136
99%       669
        ...  
52%         1
49%         1
48%         1
74%         1
2%          1
Name: count, Length: 96, dtype: int64

In [25]:
df_lodging['문의_응답률'] = df_lodging['문의_응답률'].str.rstrip('%')
df_lodging['문의_응답률'].head()

0    100
1    100
2    100
3    NaN
4    100
Name: 문의_응답률, dtype: object

In [26]:
df_lodging['문의_응답률'].isnull().sum()

36087

왜 널값일까?
-> null 값인 숙소 들어가서 응답률이 있는 곳과 비교해보기

In [27]:
# raw 테이블에서 확인해야함. 
# >> 숙소 url 값이 숙소 테이블에 없어서
df[['listing_url', 'host_response_rate']][df['host_response_rate'].isnull()]

listing_url host_response_rate
3                   https://www.airbnb.com/rooms/166370                NaN
5                    https://www.airbnb.com/rooms/40143                NaN
13                   https://www.airbnb.com/rooms/43027                NaN
19                    https://www.airbnb.com/rooms/9359                NaN
20                  https://www.airbnb.com/rooms/178188                NaN
...                                                 ...                ...
84380  https://www.airbnb.com/rooms/1112105752085360103                NaN
84383  https://www.airbnb.com/rooms/1112438676006448898                NaN
84385  https://www.airbnb.com/rooms/1112455883531799711                NaN
84390  https://www.airbnb.com/rooms/1112487387667998646                NaN
84395  https://www.airbnb.com/rooms/1112843324939236177                NaN

[36087 rows x 2 columns]

In [28]:
df[['listing_url', 'host_response_rate']][df['host_response_rate'].notnull()]

listing_url host_response_rate
0                     https://www.airbnb.com/rooms/3109               100%
1                   https://www.airbnb.com/rooms/165409               100%
2                     https://www.airbnb.com/rooms/5396               100%
4                    https://www.airbnb.com/rooms/39948               100%
6                     https://www.airbnb.com/rooms/7397               100%
...                                                 ...                ...
84391  https://www.airbnb.com/rooms/1112525964551684077                99%
84392  https://www.airbnb.com/rooms/1112562855977040914                99%
84393  https://www.airbnb.com/rooms/1112591023192990750               100%
84394  https://www.airbnb.com/rooms/1112842648117310093               100%
84396  https://www.airbnb.com/rooms/1112849968490989659               100%

[48310 rows x 2 columns]

# 검색 결과 :

- https://community.withairbnb.com/t5/Airbnb-updates/Understanding-Response-Rate-and-Acceptance-Rate/m-p/761075

Response rate(응답률)란
 - 응답률은 지난 30일 동안 24시간 이내에 귀하가 응답(수락/사전 승인 또는 거부)한 새로운 문의 및 예약 요청의 비율입니다.

 즉, host_response_rate(문의_응답률)가 null 값인 경우는 30일 이내에 새로운 문의나 예약 요청이 없었거나, 새로운 문의나 예약 요청에 응답하지 않은 것으로 해석될 수 있음.

airbnb 응답률 계산 방법 : 
 - 우리는 고객의 첫 번째 메시지나 문의 로부터 24시간 이내에만 응답을 측정합니다 . 해당 스레드의 후속 메시지는 응답률에 영향을 미치지 않습니다.

Response time(응답 시간)이란
 - 응답 시간은 지난 30일 동안 모든 새 메시지에 응답하는 데 걸린 평균 시간입니다.

host_response_time(답변_평균시간) 역시 null값인 경우가 30일 동안 새 메세지를 받지 못했거나, 받았지만 응답하지 않은 경우로 볼 수 있음.

host_acceptance_rate(예약_수락률)이란
 - 예약을 수락하거나 거부하는 빈도를 측정합니다. 게스트 문의는 수락률 계산에 포함되지 않습니다.

5/20일 월요일 팀 회의에 논의 필요해 보임. 해당 컬럼들을 어떻게 처리할 지

그나마 문의_응답률과 예약_수락률은 % 제거 후 notnull값들 뽑은 후 평균값 구해서 대체 할 수 있을 것 같음.

답변_평균시간의 경우는 범주형으로 바꾸긴 했지만, 그 이후를 잘 모르겠음.

### 일단 문의_응답률 과 예약_수락률은 평균 대체로 진행해봄.

In [29]:
df_lodging['문의_응답률']

0        100
1        100
2        100
3        NaN
4        100
        ... 
84392     99
84393    100
84394    100
84395    NaN
84396    100
Name: 문의_응답률, Length: 84397, dtype: object

In [30]:
# notnull 값들만 추출
response_rate_notnull = df_lodging['문의_응답률'][df_lodging['문의_응답률'].notnull()]
response_rate_notnull.head()

0    100
1    100
2    100
4    100
6    100
Name: 문의_응답률, dtype: object

In [31]:
# 평균값 구하기 위한 형변환 object >> int
# int인 이유 : 기존 문의_응답률에 float값이 아니여서
response_rate_notnull = response_rate_notnull.astype('int')
response_rate_notnull.head()

0    100
1    100
2    100
4    100
6    100
Name: 문의_응답률, dtype: int32

In [32]:
# 평균값 도출
# int 타입이라 round를 defalut 값으로 함.
response_rate_mean = round(response_rate_notnull.mean())
response_rate_mean

94

In [33]:
# 평균값 대체
df_lodging['문의_응답률'].fillna(response_rate_mean, inplace=True)
df_lodging['문의_응답률']
# index 3 확인 nan >> 94 적용 확인

0        100
1        100
2        100
3         94
4        100
        ... 
84392     99
84393    100
84394    100
84395     94
84396    100
Name: 문의_응답률, Length: 84397, dtype: object

In [34]:
# 예약_수락률 처리
df_lodging['예약_수락률'].unique()

array(['80%', '96%', '97%', '7%', '89%', nan, '67%', '0%', '95%', '44%',
       '40%', '100%', '10%', '71%', '48%', '88%', '50%', '69%', '86%',
       '98%', '16%', '75%', '20%', '94%', '99%', '41%', '83%', '93%',
       '29%', '90%', '21%', '13%', '57%', '91%', '33%', '92%', '32%',
       '58%', '31%', '39%', '55%', '56%', '84%', '45%', '54%', '85%',
       '78%', '76%', '60%', '87%', '18%', '2%', '72%', '22%', '37%',
       '79%', '70%', '77%', '9%', '25%', '81%', '73%', '64%', '62%',
       '34%', '27%', '6%', '82%', '63%', '19%', '24%', '35%', '74%',
       '65%', '26%', '68%', '14%', '38%', '47%', '52%', '12%', '59%',
       '43%', '53%', '36%', '23%', '15%', '49%', '3%', '66%', '61%',
       '51%', '30%', '46%', '11%', '42%', '5%', '17%', '8%', '4%', '28%',
       '1%'], dtype=object)

In [35]:
df_lodging['예약_수락률'] = df_lodging['예약_수락률'].str.rstrip('%')
df_lodging['예약_수락률'].head(6)

0     80
1     96
2     97
3      7
4     89
5    NaN
Name: 예약_수락률, dtype: object

In [36]:
# notnull 값들만 추출
acceptance_rate_notnull = df_lodging['예약_수락률'][df_lodging['예약_수락률'].notnull()]
acceptance_rate_notnull.head(6)

0    80
1    96
2    97
3     7
4    89
6    67
Name: 예약_수락률, dtype: object

In [37]:
# int로 형변환 
acceptance_rate_notnull = acceptance_rate_notnull.astype('int')
acceptance_rate_notnull.head()

0    80
1    96
2    97
3     7
4    89
Name: 예약_수락률, dtype: int32

In [38]:
# 평균값 도출
acceptance_rate_mean = round(acceptance_rate_notnull.mean())
acceptance_rate_mean

81

In [39]:
# 평균값 대체
df_lodging['예약_수락률'].fillna(acceptance_rate_mean, inplace=True)
df_lodging['예약_수락률'].head(6)
# index 5: nan >> 81 대체 확인

0    80
1    96
2    97
3     7
4    89
5    81
Name: 예약_수락률, dtype: object

In [40]:
df_lodging.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84397 entries, 0 to 84396
Data columns (total 17 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   숙소_id    84397 non-null  int64   
 1   호스트_id   84397 non-null  int64   
 2   답변_평균시간  48310 non-null  category
 3   문의_응답률   84397 non-null  object  
 4   예약_수락률   84397 non-null  object  
 5   슈퍼호스트    84306 non-null  object  
 6   숙소_수     84390 non-null  float64 
 7   숙소_지역    84397 non-null  object  
 8   숙소_유형    84397 non-null  object  
 9   수용인원수    84397 non-null  int64   
 10  욕실수      60959 non-null  float64 
 11  침실수      77480 non-null  float64 
 12  침대수      60828 non-null  float64 
 13  편의시설     84397 non-null  object  
 14  숙소_가격    60917 non-null  object  
 15  최소_숙박일   84397 non-null  int64   
 16  최대_숙박일   84397 non-null  int64   
dtypes: category(1), float64(4), int64(5), object(7)
memory usage: 10.4+ MB


이후 다른 부분부터 처리 >> 슈퍼호스트(host_is_superhost)

In [41]:
df_lodging['슈퍼호스트'].unique()

array(['f', 't', nan], dtype=object)

In [42]:
# nan 값 >> 'f' 대체 
# 'f' 값 대체 이유 : 슈퍼 호스트가 아님.
df_lodging['슈퍼호스트'].fillna('f', inplace=True)

In [43]:
df_lodging.isnull().sum()

숙소_id          0
호스트_id         0
답변_평균시간    36087
문의_응답률         0
예약_수락률         0
슈퍼호스트          0
숙소_수           7
숙소_지역          0
숙소_유형          0
수용인원수          0
욕실수        23438
침실수         6917
침대수        23569
편의시설           0
숙소_가격      23480
최소_숙박일         0
최대_숙박일         0
dtype: int64

숙소_수 확인

In [44]:
df[['listing_url', 'host_total_listings_count','host_verifications','host_identity_verified','number_of_reviews']][df['host_total_listings_count'].isnull()]

listing_url  host_total_listings_count  \
3731    https://www.airbnb.com/rooms/3986594                        NaN   
5182    https://www.airbnb.com/rooms/5872399                        NaN   
5542    https://www.airbnb.com/rooms/6300239                        NaN   
6305    https://www.airbnb.com/rooms/6875069                        NaN   
7251    https://www.airbnb.com/rooms/7447839                        NaN   
7569    https://www.airbnb.com/rooms/7526817                        NaN   
16105  https://www.airbnb.com/rooms/19262755                        NaN   

      host_verifications host_identity_verified  number_of_reviews  
3731                 NaN                    NaN                  0  
5182                 NaN                    NaN                  0  
5542                 NaN                    NaN                  0  
6305                 NaN                    NaN                  0  
7251                 NaN                    NaN                  0  
7569                 NaN                    NaN                  0  
16105                NaN                    NaN                  0

확인 결과 :
 - 7개 호스트 프로필 선택 시 전원 404 에러 발생
 - 7개 호스트 전원 신원이 검증되어 있지 않음
 - 완료한 인증 방법도 없음
 - 7개 상품 전부 후기 없음
 - 호스트의 다른 숙소 역시 페이지가 안뜸

제거 해도 될 것 같음.

In [45]:
# 숙소_수가 null값인 행 제거 
df_lodging = df_lodging.dropna(subset=['숙소_수'])

In [46]:
df_lodging.info()

<class 'pandas.core.frame.DataFrame'>
Index: 84390 entries, 0 to 84396
Data columns (total 17 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   숙소_id    84390 non-null  int64   
 1   호스트_id   84390 non-null  int64   
 2   답변_평균시간  48310 non-null  category
 3   문의_응답률   84390 non-null  object  
 4   예약_수락률   84390 non-null  object  
 5   슈퍼호스트    84390 non-null  object  
 6   숙소_수     84390 non-null  float64 
 7   숙소_지역    84390 non-null  object  
 8   숙소_유형    84390 non-null  object  
 9   수용인원수    84390 non-null  int64   
 10  욕실수      60959 non-null  float64 
 11  침실수      77474 non-null  float64 
 12  침대수      60828 non-null  float64 
 13  편의시설     84390 non-null  object  
 14  숙소_가격    60917 non-null  object  
 15  최소_숙박일   84390 non-null  int64   
 16  최대_숙박일   84390 non-null  int64   
dtypes: category(1), float64(4), int64(5), object(7)
memory usage: 11.0+ MB


84397 >> 84390

### 욕실수 , 침실수, 침대수 처리

In [47]:
# 욕실수 notnull 값들만 확인하기 
bathroom_notnull = df_lodging['욕실수'][df_lodging['욕실수'].notnull()]
bathroom_notnull

0        1.0
1        1.0
2        1.0
3        1.0
6        1.0
        ... 
84392    1.0
84393    1.0
84394    1.0
84395    1.0
84396    1.0
Name: 욕실수, Length: 60959, dtype: float64

In [48]:
bathroom_notnull.unique()

array([ 1. ,  1.5,  2. ,  2.5,  4.5,  4. ,  3. ,  3.5,  0. ,  0.5,  5. ,
        5.5,  7. ,  6. , 48. ,  6.5,  7.5,  8. , 10. ,  8.5, 30. ])

이게 뭐지?
 - 1.5 ?
 - 5.5 ? 
 - 48 ?

욕실 개수가 이상함...
소수점이 있는 숙소와 너무 많은 숙소 확인 필요

In [49]:
# df에서 listing_url + property_type + room_type  + bathrooms_text 으로 확인
# 임시로 df_temp라는 변수에 넣어서 확인
df_temp = df[['listing_url', 'property_type','room_type','bathrooms_text']][df['bathrooms']==1.5]
df_temp

listing_url       property_type  \
9                   https://www.airbnb.com/rooms/167998  Entire rental unit   
14                   https://www.airbnb.com/rooms/43055  Entire rental unit   
21                   https://www.airbnb.com/rooms/44130  Entire rental unit   
55                   https://www.airbnb.com/rooms/56662  Entire rental unit   
72                   https://www.airbnb.com/rooms/14903  Entire rental unit   
...                                                 ...                 ...   
84266  https://www.airbnb.com/rooms/1111656405283218592  Entire rental unit   
84289  https://www.airbnb.com/rooms/1112248527217605543  Entire rental unit   
84326  https://www.airbnb.com/rooms/1112301082909610836  Entire rental unit   
84337  https://www.airbnb.com/rooms/1112309632091288370  Entire rental unit   
84391  https://www.airbnb.com/rooms/1112525964551684077  Entire rental unit   

             room_type bathrooms_text  
9      Entire home/apt      1.5 baths  
14     Entire home/apt      1.5 baths  
21     Entire home/apt      1.5 baths  
55     Entire home/apt      1.5 baths  
72     Entire home/apt      1.5 baths  
...                ...            ...  
84266  Entire home/apt      1.5 baths  
84289  Entire home/apt      1.5 baths  
84326  Entire home/apt      1.5 baths  
84337  Entire home/apt      1.5 baths  
84391  Entire home/apt      1.5 baths  

[5145 rows x 4 columns]

In [50]:
df_temp['room_type'].value_counts()

room_type
Entire home/apt    4600
Private room        526
Shared room          17
Hotel room            2
Name: count, dtype: int64

In [51]:
df_temp = df[['listing_url', 'property_type','room_type','bathrooms_text']][df['bathrooms']==0]
df_temp

listing_url  \
749                 https://www.airbnb.com/rooms/726886   
1170               https://www.airbnb.com/rooms/1179811   
1413               https://www.airbnb.com/rooms/1385662   
1894               https://www.airbnb.com/rooms/2048626   
2095               https://www.airbnb.com/rooms/2289322   
...                                                 ...   
83669  https://www.airbnb.com/rooms/1109202511119628786   
83885  https://www.airbnb.com/rooms/1110667965594810434   
84125  https://www.airbnb.com/rooms/1110651899288071181   
84243  https://www.airbnb.com/rooms/1112206680420925410   
84280  https://www.airbnb.com/rooms/1112236160393702684   

                     property_type        room_type  bathrooms_text  
749             Entire rental unit  Entire home/apt         0 baths  
1170            Entire rental unit  Entire home/apt         0 baths  
1413   Private room in rental unit     Private room  0 shared baths  
1894            Entire rental unit  Entire home/apt         0 baths  
2095            Entire rental unit  Entire home/apt         0 baths  
...                            ...              ...             ...  
83669  Private room in rental unit     Private room  0 shared baths  
83885  Private room in rental unit     Private room  0 shared baths  
84125           Entire rental unit  Entire home/apt         0 baths  
84243           Entire rental unit  Entire home/apt         0 baths  
84280           Entire rental unit  Entire home/apt         0 baths  

[279 rows x 4 columns]

In [52]:
df_temp['room_type'].value_counts()

room_type
Entire home/apt    239
Private room        34
Shared room          4
Hotel room           2
Name: count, dtype: int64

## 검색 결과

유럽 표준 EN 12056에서 제시하는 화장실 기준

1. 전체 화장실(Full bathroom):
- 변기, 세면대, 욕조 또는 샤워기가 모두 포함된 완전한 화장실
- 최소 면적 4㎡ 이상 확보
- 이를 사람들은 1로 표기함.

2. 부분 화장실(Half bathroom):
- 변기와 세면대만 포함된 화장실
- 최소 면적 2㎡ 이상 확보
- 이를 사람들은 0.5라 표기함

따라서 1.5라는 표현은 전체 화장실 1개와 부분 화장실 1개라는 의미로 해석할 수 있음

#### 그러면 30개, 48개는??

In [53]:
df_temp = df[['listing_url', 'property_type','room_type','bathrooms_text']][df['bathrooms']==30]
df_temp

listing_url  property_type  \
76331  https://www.airbnb.com/rooms/1070960072975209098  Room in hotel   
76378  https://www.airbnb.com/rooms/1071020436354803291  Room in hotel   

          room_type bathrooms_text  
76331  Private room       30 baths  
76378  Private room       30 baths

확인 결과 : 30개 = 호텔 전체 욕실 수 같음

-> 30개의 방이 존재 -> 각 방에 1개씩 완전한 화장실이 있음

In [54]:
df_temp = df[['listing_url', 'property_type','room_type','bathrooms_text']][df['bathrooms']==48]
df_temp

listing_url           property_type  \
17247  https://www.airbnb.com/rooms/20433587  Room in boutique hotel   

          room_type bathrooms_text  
17247  Private room       48 baths

확인 결과 : 48개짜리 숙소는 페이지가 없는지 airbnb 메인 홈으로 들어와짐.

### 5 = Master bathroom 이라는 뜻으로 좀 다르다는 정보를 확인

Master bathroom = 마스터 침실에 인접한 전용 욕실
 - 변기 + 세면대 2개 + 욕조 + 샤워부스

5인 값만 확인

In [55]:
df_temp = df[['listing_url', 'property_type','room_type','bathrooms_text']][df['bathrooms']==5]
df_temp

listing_url  \
2806               https://www.airbnb.com/rooms/3087414   
3311               https://www.airbnb.com/rooms/3430844   
9509              https://www.airbnb.com/rooms/10480504   
16252             https://www.airbnb.com/rooms/19501310   
19662             https://www.airbnb.com/rooms/24387710   
29728             https://www.airbnb.com/rooms/39851605   
30964             https://www.airbnb.com/rooms/41845069   
32072             https://www.airbnb.com/rooms/43339272   
36163             https://www.airbnb.com/rooms/50773342   
38739             https://www.airbnb.com/rooms/54066057   
39208   https://www.airbnb.com/rooms/551975476695827301   
41780   https://www.airbnb.com/rooms/628279299800770288   
43900   https://www.airbnb.com/rooms/665990040554168884   
45600   https://www.airbnb.com/rooms/697150397271935389   
46515   https://www.airbnb.com/rooms/725945295444337513   
47786   https://www.airbnb.com/rooms/754885427025620893   
52176   https://www.airbnb.com/rooms/847653221708092409   
58959   https://www.airbnb.com/rooms/927468778094017922   
59923   https://www.airbnb.com/rooms/933978012684095099   
62435   https://www.airbnb.com/rooms/958477038534616358   
65111   https://www.airbnb.com/rooms/983188549919323614   
65442   https://www.airbnb.com/rooms/986180692294457272   
65697   https://www.airbnb.com/rooms/987598507410314671   
69449  https://www.airbnb.com/rooms/1022236264230180521   
76078  https://www.airbnb.com/rooms/1069528345792695797   
77794  https://www.airbnb.com/rooms/1077543955821634038   
79136  https://www.airbnb.com/rooms/1085265558004121790   
79332  https://www.airbnb.com/rooms/1086055288477640741   
80477  https://www.airbnb.com/rooms/1092720121063377343   
82302  https://www.airbnb.com/rooms/1101310060020360515   
83299  https://www.airbnb.com/rooms/1106346421059497241   
83580  https://www.airbnb.com/rooms/1107894589115893027   

                           property_type        room_type bathrooms_text  
2806                  Entire rental unit  Entire home/apt        5 baths  
3311                         Entire home  Entire home/apt        5 baths  
9509                  Entire rental unit  Entire home/apt        5 baths  
16252                   Entire townhouse  Entire home/apt        5 baths  
19662                 Room in aparthotel     Private room        5 baths  
29728                 Entire rental unit  Entire home/apt        5 baths  
30964                        Entire home  Entire home/apt        5 baths  
32072                 Entire rental unit  Entire home/apt        5 baths  
36163                       Entire condo  Entire home/apt        5 baths  
38739                       Entire villa  Entire home/apt        5 baths  
39208                 Entire rental unit  Entire home/apt        5 baths  
41780  Private room in bed and breakfast     Private room        5 baths  
43900               Entire vacation home  Entire home/apt        5 baths  
45600                        Entire home  Entire home/apt        5 baths  
46515                 Entire rental unit  Entire home/apt        5 baths  
47786                 Entire rental unit  Entire home/apt        5 baths  
52176                 Entire rental unit  Entire home/apt        5 baths  
58959                   Entire townhouse  Entire home/apt        5 baths  
59923        Private room in rental unit     Private room        5 baths  
62435                       Entire villa  Entire home/apt        5 baths  
65111                        Entire home  Entire home/apt        5 baths  
65442                 Entire rental unit  Entire home/apt        5 baths  
65697                        Entire home  Entire home/apt        5 baths  
69449                               Boat  Entire home/apt        5 baths  
76078                 Entire rental unit  Entire home/apt        5 baths  
77794                 Entire rental unit  Entire home/apt        5 baths  
79136                 Entire rental unit  Entire home/apt        5 bath

In [56]:
df_temp.count()

listing_url       32
property_type     32
room_type         32
bathrooms_text    32
dtype: int64

32개 확인 결과 : 그저 욕실 5개인 것으로 확인함.

최빈값으로 대체
 - 평균값과 중위값 확인 후 진행

In [57]:
df_lodging['욕실수'].describe()

count    60959.000000
mean         1.208763
std          0.571918
min          0.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         48.000000
Name: 욕실수, dtype: float64

In [58]:
bathroom_notnull.value_counts()

욕실수
1.0     47458
1.5      5145
2.0      5094
2.5      1252
3.0       931
0.0       279
0.5       270
3.5       246
4.0       143
4.5        61
5.0        32
5.5        13
6.0        12
6.5         7
7.0         6
7.5         3
8.0         2
30.0        2
48.0        1
10.0        1
8.5         1
Name: count, dtype: int64

In [59]:
bathroom_notnull.describe()

count    60959.000000
mean         1.208763
std          0.571918
min          0.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         48.000000
Name: 욕실수, dtype: float64

In [60]:
# 욕실 수의 null값은 가장 많은 1.0으로 대체하겠음.
# 이유 : 평균값 또는 중위값 확인 결과 1.0에 가깝기 때문.
df_lodging['욕실수'].fillna(1.0, inplace=True)

In [61]:
df_lodging['욕실수'].head()

0    1.0
1    1.0
2    1.0
3    1.0
4    1.0
Name: 욕실수, dtype: float64

In [62]:
df_lodging['욕실수'].describe()

count    84390.000000
mean         1.150800
std          0.494988
min          0.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         48.000000
Name: 욕실수, dtype: float64

1.0으로 채우면서 std(표준편차)값이 떨어짐.
- 문제 발생할지 모르겠음.
- 일단 진행

### 침실수
 - 욕실수와 마찬가지로 평균값과 중위값 확인 후 진행하겠음
 - 이유 : 욕실수와 비슷하기 때문에

In [63]:
df_lodging['침실수'].unique()

array([ 1.,  0., nan,  2.,  3.,  6.,  4.,  5.,  7., 16., 50., 32., 12.,
        8., 38., 14.,  9., 10., 33., 11., 41., 30.])

In [64]:
df_lodging['침실수'].value_counts()

침실수
1.0     48606
2.0     15269
0.0      6274
3.0      5512
4.0      1407
5.0       285
6.0        75
7.0        26
8.0         5
33.0        2
30.0        2
16.0        2
12.0        1
50.0        1
38.0        1
14.0        1
9.0         1
10.0        1
11.0        1
41.0        1
32.0        1
Name: count, dtype: int64

In [65]:
# 숫자가 큰 이유 확인
df_temp = df[['listing_url', 'property_type','room_type']][df['bedrooms'] >= 10 ]
df_temp

listing_url  \
7256               https://www.airbnb.com/rooms/7449162   
17247             https://www.airbnb.com/rooms/20433587   
17673             https://www.airbnb.com/rooms/21216654   
18755             https://www.airbnb.com/rooms/22925880   
19927             https://www.airbnb.com/rooms/24667815   
20988             https://www.airbnb.com/rooms/26253224   
27189             https://www.airbnb.com/rooms/36124445   
31955             https://www.airbnb.com/rooms/43084931   
32106             https://www.airbnb.com/rooms/43369014   
54966   https://www.airbnb.com/rooms/888128066633316596   
57359   https://www.airbnb.com/rooms/913477855231933260   
63771   https://www.airbnb.com/rooms/970856355633037590   
76331  https://www.airbnb.com/rooms/1070960072975209098   
76378  https://www.airbnb.com/rooms/1071020436354803291   

                property_type        room_type  
7256       Entire rental unit  Entire home/apt  
17247  Room in boutique hotel     Private room  
17673      Entire rental unit  Entire home/apt  
18755      Entire rental unit  Entire home/apt  
19927  Room in boutique hotel       Hotel room  
20988           Room in hotel     Private room  
27189   Private room in condo     Private room  
31955           Room in hotel     Private room  
32106           Room in hotel     Private room  
54966      Entire rental unit  Entire home/apt  
57359           Room in hotel     Private room  
63771             Entire home  Entire home/apt  
76331           Room in hotel     Private room  
76378           Room in hotel     Private room

확인 결과 : 전체 수를 적은 경우임.
 - 예를 들어 호텔의 방이 41개면 침실을 41개로 적어놓은 것들임.

In [66]:
df_lodging['침실수'].describe()

count    77474.000000
mean         1.339546
std          0.922877
min          0.000000
25%          1.000000
50%          1.000000
75%          2.000000
max         50.000000
Name: 침실수, dtype: float64

In [67]:
# 1.0으로 대체
df_lodging['침실수'].fillna(1.0, inplace=True)
df_lodging['침실수']

0        1.0
1        0.0
2        0.0
3        1.0
4        1.0
        ... 
84392    1.0
84393    1.0
84394    1.0
84395    1.0
84396    0.0
Name: 침실수, Length: 84390, dtype: float64

In [68]:
df_lodging['침실수'].describe()

count    84390.000000
mean         1.311719
std          0.889143
min          0.000000
25%          1.000000
50%          1.000000
75%          2.000000
max         50.000000
Name: 침실수, dtype: float64

침대수
 - 이게 애매함...
 - 침실수 확인하면서 어느정도 본 결과 1침실 2침대인 경우가 존재함..
 - 예를 들어 싱글베드 2개...

In [69]:
df_lodging['침대수'].unique()

array([ 1., nan,  2.,  3.,  5.,  9.,  0.,  7.,  4.,  8.,  6., 10., 11.,
       13., 17., 77., 14., 12., 30.])

In [70]:
df_lodging['침대수'].value_counts()
# 최빈값과 그 다음 값이 약 1만 차이

침대수
1.0     27020
2.0     17620
3.0      6961
0.0      4169
4.0      3155
5.0      1128
6.0       451
7.0       175
8.0        87
9.0        30
10.0       13
11.0        7
12.0        5
13.0        2
30.0        2
17.0        1
77.0        1
14.0        1
Name: count, dtype: int64

In [71]:
df_lodging['침대수'].describe()
# 평균값은 1.7로 2에 가깝고
# 중위값은 1.0임

count    60828.000000
mean         1.755129
std          1.254172
min          0.000000
25%          1.000000
50%          1.000000
75%          2.000000
max         77.000000
Name: 침대수, dtype: float64

In [72]:
df_lodging['침대수'].isnull().sum()
# null 값이 2만개를 넘어가서
# 뭘로 대체하느냐에 따라 최빈값이 바뀔 우려가 있음.

23562

우선 침실에 침대 1개 이상은 존재하므로 1.0으로 대체해서 진행하겠음.
 - 차후 문제 생길 시 평균값으로 대체하겠음 (1.75)

In [73]:
# 1.0으로 대체
df_lodging['침대수'].fillna(1.0, inplace=True)
df_lodging['침대수']

0        1.0
1        1.0
2        1.0
3        1.0
4        1.0
        ... 
84392    1.0
84393    1.0
84394    2.0
84395    2.0
84396    1.0
Name: 침대수, Length: 84390, dtype: float64

In [74]:
df_lodging['침대수'].describe()
# 표준편차(std)는 0.1 정도만 줄어듬

count    84390.000000
mean         1.544294
std          1.117375
min          0.000000
25%          1.000000
50%          1.000000
75%          2.000000
max         77.000000
Name: 침대수, dtype: float64

In [75]:
df_lodging.isnull().sum()

숙소_id          0
호스트_id         0
답변_평균시간    36080
문의_응답률         0
예약_수락률         0
슈퍼호스트          0
숙소_수           0
숙소_지역          0
숙소_유형          0
수용인원수          0
욕실수            0
침실수            0
침대수            0
편의시설           0
숙소_가격      23473
최소_숙박일         0
최대_숙박일         0
dtype: int64

숙소_가격 결측치
 - 23473개

In [76]:
df_lodging['숙소_가격']

0        $250.00
1        $140.00
2        $107.00
3        $200.00
4            NaN
          ...   
84392     $80.00
84393    $167.00
84394     $98.00
84395    $296.00
84396     $71.00
Name: 숙소_가격, Length: 84390, dtype: object

In [77]:
# $ 제거
df_lodging['숙소_가격'] = df_lodging['숙소_가격'].str.lstrip('$')
df_lodging['숙소_가격'][:3]

0    250.00
1    140.00
2    107.00
Name: 숙소_가격, dtype: object

In [78]:
# 숙소 가격에서 쉼표(,) 제거
df_lodging['숙소_가격'] = df_lodging['숙소_가격'].str.replace(',', '')
df_lodging['숙소_가격'][:3]

0    250.00
1    140.00
2    107.00
Name: 숙소_가격, dtype: object

In [79]:
# null값 제외, 나머지 값들만 추출
lodging_price_notnull = df_lodging['숙소_가격'][df_lodging['숙소_가격'].notnull()]
lodging_price_notnull

0        250.00
1        140.00
2        107.00
3        200.00
6        134.00
          ...  
84392     80.00
84393    167.00
84394     98.00
84395    296.00
84396     71.00
Name: 숙소_가격, Length: 60917, dtype: object

In [80]:
# 형변환 :  object >> float
lodging_price_notnull = lodging_price_notnull.astype('float')
lodging_price_notnull[:3]

0    250.0
1    140.0
2    107.0
Name: 숙소_가격, dtype: float64

In [81]:
mean_price = round(lodging_price_notnull.mean(),2)
mean_price

288.66

In [82]:
# 평균값 대체
df_lodging['숙소_가격'].fillna(mean_price, inplace=True)
df_lodging['숙소_가격']
# index 4 : nan >> 288.66 확인

0        250.00
1        140.00
2        107.00
3        200.00
4        288.66
          ...  
84392     80.00
84393    167.00
84394     98.00
84395    296.00
84396     71.00
Name: 숙소_가격, Length: 84390, dtype: object

In [83]:
df_lodging.head()

숙소_id  호스트_id             답변_평균시간 문의_응답률 예약_수락률 슈퍼호스트  숙소_수  \
0    3109    3631        within a day    100     80     f   2.0   
1  165409  788732  within a few hours    100     96     t   1.0   
2    5396    7903      within an hour    100     97     f   3.0   
3  166370  793032                 NaN     94      7     f   3.0   
4   39948  171483        within a day    100     89     t   2.0   

            숙소_지역            숙소_유형  수용인원수  욕실수  침실수  침대수  \
0    Observatoire  Entire home/apt      2  1.0  1.0  1.0   
1          Temple  Entire home/apt      2  1.0  0.0  1.0   
2  Hôtel-de-Ville  Entire home/apt      2  1.0  0.0  1.0   
3        Entrepôt  Entire home/apt      4  1.0  1.0  1.0   
4          Bourse     Private room      2  1.0  1.0  1.0   

                                                편의시설   숙소_가격  최소_숙박일  최대_숙박일  
0  ["Cooking basics", "Luggage dropoff allowed", ...  250.00       2      30  
1  ["Lockbox", "Wifi", "Kitchen", "Refrigerator",...  140.00       3     365  
2  ["Cooking basics", "Long term stays allowed", ...  107.00       1    1125  
3  ["Shampoo", "Kitchen", "Washer", "Cooking basi...  200.00       2     120  
4  ["Extra pillows and blankets", "Wifi", "Privat...  288.66       5      20

In [84]:
df_lodging.isnull().sum()

숙소_id          0
호스트_id         0
답변_평균시간    36080
문의_응답률         0
예약_수락률         0
슈퍼호스트          0
숙소_수           0
숙소_지역          0
숙소_유형          0
수용인원수          0
욕실수            0
침실수            0
침대수            0
편의시설           0
숙소_가격          0
최소_숙박일         0
최대_숙박일         0
dtype: int64

최대/최소 숙박일 확인

In [85]:
df_lodging['최대_숙박일'].unique()

array([        30,        365,       1125,        120,         20,
              130,         50,        730,          6,        300,
                7,         90,         21,        360,         60,
               16,         31,         40,         25,        305,
              600,         29,        550,         27,         70,
              180,        140,        372,         15,        190,
             1000,         14,        500,        572,        400,
                5,        150,         10,         28,       1124,
              100,        750,          4,         13,         11,
                8,        999,        306,        368,         45,
              700,        200,        240,         36,         12,
                3,         42,          2,         17,         24,
               58,       1001,         35,        119,         18,
                9,        350,         62,        800,        125,
               22,         80,        450,        380,        

- 2147483647 ???
- 1000 ???
- 3000 ???

이게 뭔...

In [86]:
# 빈도수 확인
df_lodging['최대_숙박일'].value_counts()
# 1125일 뭐길래 이렇게 많은지 확인 필요 

최대_숙박일
1125    28890
365     23033
30       5502
15       2208
120      2193
        ...  
308         1
1111        1
312         1
620         1
239         1
Name: count, Length: 261, dtype: int64

In [87]:
df_lodging['최대_숙박일'].describe()

count    8.439000e+04
mean     2.608600e+04
std      7.392459e+06
min      1.000000e+00
25%      4.200000e+01
50%      3.650000e+02
75%      1.125000e+03
max      2.147484e+09
Name: 최대_숙박일, dtype: float64

In [88]:
df_lodging['최대_숙박일'].max()

2147483647

In [89]:
df_temp = df[['listing_url', 'property_type','room_type','price','minimum_nights']][df['maximum_nights']==2147483647]
df_temp

listing_url       property_type  \
1145  https://www.airbnb.com/rooms/1088383  Entire rental unit   

            room_type    price  minimum_nights  
1145  Entire home/apt  $146.00               3

In [98]:
# 최소_숙박일
df_lodging['최소_숙박일'].unique()

array([   2,    3,    1,    5,  365,   10,   20,    4,   12,    7,   90,
         30,  180,    6,  250,  121,   21,   15,    8,   13,   60,   19,
         28,   14,   27,  500,   11,   25,  270,  120,   91,   35,   18,
          9,   50,  210, 1000,  400,   80,   92,   31,   45,   29,  100,
        300,   40,   89,   62,   24,   32,  260,  140,  360,   79,  150,
         54,   23,  160,   16,  110,   22,  168,  999,  366,  370,  112,
        252,   55,   17,   93,   59,   39,   95,   53,   65,  199,  200,
         57,  185,  170,  275,  274,   26,   70,   99,  375,   75,  730,
         33, 1123,   36,  175,  178,   42,  115,  130,  350,   85,  280,
         78,   84,  303], dtype=int64)

여기도 이상하네..
 - 1000 ???
 - 999 ?? 
 - 1123 ??

In [100]:
df_lodging['최소_숙박일'].value_counts()
# 일단 최소 1년 짜리도 많으므로 확인

최소_숙박일
1      18864
2      16754
3      15129
365     9765
4       5837
       ...  
53         1
199        1
57         1
185        1
303        1
Name: count, Length: 102, dtype: int64

In [109]:
df_temp = df[['listing_url', 'property_type','room_type','price','minimum_nights']][df['minimum_nights']==199]
df_temp

listing_url       property_type  \
21522  https://www.airbnb.com/rooms/27106571  Entire rental unit   

             room_type price  minimum_nights  
21522  Entire home/apt   NaN             199

In [105]:
df_temp['room_type'].value_counts()

room_type
Entire home/apt    9680
Hotel room           66
Private room         24
Shared room           2
Name: count, dtype: int64

확인 해보니 그저 평범한 상품이었음...
 - 최대_숙박일에서의 이상치들은 그저 게스트가 돈이 되는대로 가능하게 해놓은 것 같음.
 - 최소_숙박일에서의 이상치들은 대부분이 부동산 계약이라고 보면 될 것 같음.
    - 아파트 1년 계약이 대부분이나 그 이상 역시 비슷한 계약임. 다만, 계약일수가 더 길뿐임.
    - 이외에 53, 199... 등등 역시 그저 장기 임대라고 볼 수 있음.
    
처리해야하나??
 - 일단 넘어가보기

리뷰 테이블 결측치 처리

In [90]:
df_review.head()

리뷰수  30일_리뷰수  리뷰점수  숙소_정확성_리뷰점수  숙소_청결도_리뷰점수  숙소_체크인_리뷰점수  숙소_소통_리뷰점수  \
0    4        0  5.00         5.00         5.00         5.00        5.00   
1  106        0  4.68         4.72         4.59         4.87        4.89   
2  393        5  4.60         4.63         4.57         4.81        4.84   
3   78        0  4.56         4.74         4.37         4.82        4.79   
4  126        3  4.86         4.93         4.82         4.96        4.97   

   숙소_위치_리뷰점수  숙소_가격_리뷰점수  평균_리뷰수  
0        5.00        5.00    0.05  
1        4.95        4.74    0.69  
2        4.96        4.58    2.19  
3        4.65        4.48    0.51  
4        4.94        4.82    0.99

In [91]:
df_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84397 entries, 0 to 84396
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   리뷰수          84397 non-null  int64  
 1   30일_리뷰수      84397 non-null  int64  
 2   리뷰점수         58827 non-null  float64
 3   숙소_정확성_리뷰점수  58793 non-null  float64
 4   숙소_청결도_리뷰점수  58797 non-null  float64
 5   숙소_체크인_리뷰점수  58784 non-null  float64
 6   숙소_소통_리뷰점수   58795 non-null  float64
 7   숙소_위치_리뷰점수   58787 non-null  float64
 8   숙소_가격_리뷰점수   58782 non-null  float64
 9   평균_리뷰수       58822 non-null  float64
dtypes: float64(8), int64(2)
memory usage: 6.4 MB


In [92]:
df_review.isnull().sum()

리뷰수                0
30일_리뷰수            0
리뷰점수           25570
숙소_정확성_리뷰점수    25604
숙소_청결도_리뷰점수    25600
숙소_체크인_리뷰점수    25613
숙소_소통_리뷰점수     25602
숙소_위치_리뷰점수     25610
숙소_가격_리뷰점수     25615
평균_리뷰수         25575
dtype: int64

In [96]:
df_review['리뷰수'].value_counts()

리뷰수
0      25576
1       6463
2       4449
3       3651
4       2948
       ...  
626        1
466        1
401        1
517        1
593        1
Name: count, Length: 561, dtype: int64

In [93]:
df_review['리뷰점수'].value_counts()

리뷰점수
5.00    17058
4.50     2155
4.00     2097
4.67     1854
4.75     1532
        ...  
3.54        1
2.20        1
3.59        1
3.36        1
2.40        1
Name: count, Length: 183, dtype: int64

In [94]:
# 리뷰 점수가 null값 중 리뷰수의 개수
df_review.loc[df_review['리뷰점수'].isnull(), '리뷰수'].value_counts()
# 리뷰점수가 null인 값은 리뷰도 없음

리뷰수
0    25570
Name: count, dtype: int64

즉, 리뷰수 중 0 인 25576개 중 25570은 리뷰점수도 없는 값임.

제거해도 될 것 같음
 - 제거 시 숙소 테이블과 결합 불가 >> 행 길이가 달라서

숙소 테이블 처리 전에 합친 테이블로 다시 정리필요.

탐색적 데이터 분석(EDA)